### ISSUE TITLE
BUG "redundant" transcript

### ISSUE THREAD

**Issue by swarbred** [EI-CoreBioinformatics/mikado#267]
@lucventurini 
@gemygk 

Normal transcript run i.e. only_reference_update = false

mikado.Calendula_officinalis_EIV1.1_Contig_0000004G1.3 contains mikado.Calendula_officinalis_EIV1.1_Contig_0000004G1.2

<img width="862" alt="Screen Shot 2020-02-11 at 17 10 40" src="https://user-images.githubusercontent.com/3267809/74261922-e15db480-4cf3-11ea-8fd2-a2ab2bea1f9e.png">

see mikado compare --internal output this comes up as a c relationship

[out_swarbred.mikado-2.0rc6_3dc2ed9_CBG.configuration_run2.25701404.err.txt](https://github.com/EI-CoreBioinformatics/mikado/files/4187138/out_swarbred.mikado-2.0rc6_3dc2ed9_CBG.configuration_run2.25701404.err.txt)
[mikado_compare.tmap.txt](https://github.com/EI-CoreBioinformatics/mikado/files/4187143/mikado_compare.tmap.txt)
[mikado.loci.gff3.txt](https://github.com/EI-CoreBioinformatics/mikado/files/4187146/mikado.loci.gff3.txt)

Note this is the only occurrence i've seen so perhaps some form of edge case

**Comment by lucventurini** [comment]
Hi @swarbred , this is a very probable duplicate of #266 reported by @gemygk, that bug depends by mikado not re-evaluating transcripts properly for their class code after the padding.
Could you please try to redo the analysis with the updates in `issue-266`? Hopefully those should solve the issue.

**Comment by swarbred** [comment]
@lucventurini 

installed bf4976c unfortunately this doesn't resolve the issue. I have one change in the transcript selected between the two versions but the same two "redundant" transcripts are present with the same structure as shown in the above image.

**Comment by lucventurini** [comment]
Hi @swarbred , OK, thanks for testing.
I am revising the log now, hopefully I will get an idea of what's going on soon.

**Comment by swarbred** [comment]
Just to add also now confirmed that I'm seeing multiple occurrences of these redundant models (so not such an edge case)

![Screen Shot 2020-02-12 at 10 57 39](https://user-images.githubusercontent.com/3267809/74328832-8923c280-4d86-11ea-8a4a-3d1ed41ee38c.png)

**Comment by lucventurini** [comment]
PS: could you please attach here the log and GFF for the second run?

**Comment by lucventurini** [comment]
> Just to add also now confirmed that I'm seeing multiple occurrences of these redundant models (so not such an edge case)

Mmm there definitely seems to be a bug in detecting duplicates after padding. I will have to investigate the code in that area.

**Comment by lucventurini** [comment]
For the time being I cannot replicate the issue on my end. I would need access to the files, I think, or find a way to create a test case on my end.

I think I know what is happening though. My understanding (using 1, 2 etc to refer to the transcripts, using how Mikado has named them):

- Transcript 1 creates the locus. As we are doing the padding, **redundant class codes `n`, `=` and `_` are valid for the time being**.
- Transcript 2, 3, 4 and 5 are added; at this point, none is redundant.
- Transcript 2 is consistently marked as being the second-best in terms of scoring. **This is crucial**.
- Start of the padding procedure. **Due to limitations in the number of splice sites that can be bridged and/or maximum expansion, transcript 2 is selected to be expanded only on the 5'**.
- At this point, transcript 2 is the second-best in terms of scoring, but is redundant with transcript 3.
- Creation of the new locus with the padded transcripts. Again, accept redundant transcripts. The order is:
  - 1: no controversy
  - 2: added, it's valid because it's a `j`.
  - 3: normally, *it would be invalid* because it's a `n` (perfect expansion of no. 2). But we are adding it anyway, see above.
  - 4 and 5.
- Trying to remove "redundant" transcript now. But this only takes into account perfectly identical transcripts; so both no. 2 and no. 3 remain in.

I do not have a perfect solution for this. There are three issues that I see:
- I am not persuaded that we should keep in the limitations on splice sites and maximum length. Without those, no. 2 would have been expanded to the end, as well as all the others (including the primary transcript). After the expansion, Mikado should have picked up that no. 2 and 3 were redundant.
- (**this is quite critical**) the order of insertion matters. In this specific case, the fact that no. 2 was inserted first mean that it would not ever have been marked as redundant.
- the procedure with which I check the redundancy is quite convoluted. I should definitely try to make it lighter.

**Comment by swarbred** [comment]
HI @lucventurini 

Agree with your comment on why this is not being detected and removed, I've just reran with ts_distance = 20000
ts_max_splices = 20

and the redundant transcript is now removed

```
2020-02-12 18:20:35,682 - Calendula_officinalis_EIV1.1_Contig_0000004:1908-10140 - locus.py:388 - INFO - __remove_redundant_after_padding - LociProcesser-1 - Removing CoRay.hq_lq_UnnamedSample_HQ_transcript/36743.mrna1 from locus gene:Calendula_officinalis_EIV1.1_Contig_0000004+:1940-9027 because after padding it is redundant with CoRay.hq_lq_UnnamedSample_HQ_transcript/11808.mrna1
```

>>>I am not persuaded that we should keep in the limitations on splice sites and maximum length. Without those, no. 2 would have been expanded to the end, as well as all the others (including the primary transcript). After the expansion, Mikado should have picked up that no. 2 and 3 were redundant.

I would be wary of removing these, there would seem to be a danger that the primary and higher scoring transcripts could end up being padded to resemble a transcript that scored much lower. The ts limits help limit how much the model can change form the original transcript. In certain use cases we make use of the label scoring as specified in the list.txt file (i'm doing that as a way of ensuring reference proteins are chosen as the primary model) I will therefore run with "min_score_perc =" set very low. Having no ts limits risks extending models in ways that are undesirable.

We do need a solution though as we can have these type of redundant models in the output.

I'm going to run the full set of sequences with 
```
keep_cds_disrupted_by_ri = false
keep_retained_introns = true
max_isoforms = 15
min_cdna_overlap = 0.4
min_cds_overlap = 0.4
min_score_perc = 0.4
only_confirmed_introns = true
pad = true
redundant_ccodes = ["c", "m", "_", "=", "n"]
report = true
ts_distance = 20000
ts_max_splices = 20
valid_ccodes = ["j", "J", "G", "h"]
```
just to see how this looks in a normal run.

**Comment by lucventurini** [comment]
Hi @swarbred , @gemygk 
I should have tackled point 2 in abc1c6f. 
Now there is a unit test that confirms that the case we have above **should** be removed.

If you could install and test with this commit, and @gemygk also check about the class codes being correct (see #266), we might solve two issues with one commit :-)

**Comment by swarbred** [comment]
@lucventurini installed abc1c6f

No joy i'm afraid the contained transcript is still present.


[mikado.loci.gff3.txt](https://github.com/EI-CoreBioinformatics/mikado/files/4201751/mikado.loci.gff3.txt)
[mikado_compare.tmap.txt](https://github.com/EI-CoreBioinformatics/mikado/files/4201755/mikado_compare.tmap.txt)
[out_swarbred.mikado-2.0rc6_abc1c6f_CBG.configuration_run2.25738722.err.txt](https://github.com/EI-CoreBioinformatics/mikado/files/4201756/out_swarbred.mikado-2.0rc6_abc1c6f_CBG.configuration_run2.25738722.err.txt)

**Comment by lucventurini** [comment]
Hi @swarbred,
OK, I am revising the whole procedure. Stay tuned, I should have updated code soon.

**Comment by lucventurini** [comment]
Hi @swarbred,
please see 0cfb20ad.
**WARNING: this changes quite a bit under the hood**. The results should be guaranteed to be the same (and indeed unit tests passed without a glitch), but please watch out when testing.

**Comment by swarbred** [comment]
Hi @lucventurini 

So first hurdle passed my test data for this issue has the required ccode in the tmap file. We end up with a slightly different set of transcripts than the original run but nothing I would object to.

I will do the full run and then see how that compares.

**Comment by lucventurini** [comment]
Hi @swarbred , great news.
@gemygk if you could also please check that this solves #266 that would be great.

Also, how are we with #255? If you can confirm that these three issues are all solved I would try to merge them all back to `master`. I am experimenting with that at the moment and the changes should be compatible, but I'd rather start pull things together before the various issue-branches diverge further.

**Comment by lucventurini** [comment]
PS: also, I know that's a different test, but if you could check #263 that would be great. I think the latest changes in that branch should solve the issue, but I would be needing your confirmation.

**Comment by swarbred** [comment]
@luca

Running 1258c37 (which I assume also has this redundant fix)

<img width="898" alt="Screen Shot 2020-02-19 at 08 43 39" src="https://user-images.githubusercontent.com/3267809/74817522-effc2b00-52f4-11ea-813c-06fb84940762.png">

The model (.4) highlighted in red has a redundant exon structure (to the .1) but differs in the CDS the same is true for .2 and .3

I assume the logic currently takes into account the CDS structure as well as the exon structure. All three models are class code = .

The difference in start site is related to max_regression = 0.1
Looking at the .3 model it would seem to just be within the 10% I will test this but it requires a serialise rerun.

There is an argument for taking into account the CDS but it does leave these types of examples which seem a little odd and these models should probably be regarded as redundant, thoughts?

[mikado.loci.gff3.txt](https://github.com/EI-CoreBioinformatics/mikado/files/4224158/mikado.loci.gff3.txt)
[pick.log.txt](https://github.com/EI-CoreBioinformatics/mikado/files/4224159/pick.log.txt)

**Comment by swarbred** [comment]
@lucventurini my other query would be why the 3' end was not padded (it's easily within the padding extension requirements), if the 5' end cannot be successfully padded (because the orf would change) is the 3' padding also rejected?

**Comment by lucventurini** [comment]
@swarbred 

> Running 1258c37 (which I assume also has this redundant fix)

No, these are two different branches. The changes for this issue are in `issue-266`, so the last commit would be 0cfb20a. I can create a `develop` branch with all the changes of the three issues branches merged together, if necessary.

> There is an argument for taking into account the CDS but it does leave these types of examples which seem a little odd and these models should probably be regarded as redundant, thoughts?

I need to double check whether mikado does consider the CDS. I see your point and I will have a thought about it.

> my other query would be why the 3' end was not padded (it's easily within the padding extension requirements), if the 5' end cannot be successfully padded (because the orf would change) is the 3' padding also rejected?

The padding at both ends happens in a single go, not separately. I will have to double check how failures like these are handled, I see your point on this.

**Comment by swarbred** [comment]
Actually this is 0cfb20a NOT 1258c37 but the results of the test run on the region i.e. the files I attached are not the same as the full run (the pic) will check (sorry messed up the gene ids so correected the above comment)

**Comment by swarbred** [comment]
@lucventurini 
Well I cant reproduce this when running on  the test region with 0cfb20a the logs (as attached) show this is being handled correctly e.g.

```
CoDisc_scallop_CoDisc_sca.427.0.2 is not a valid splicing isoform. Ccode: =
```
The output models are as expected
I've checked and the run shown in the pic above is 0cfb20a with the same config as the test so the only variation is the number of threads and -lv mode.

File as odd for now as it seems at least on the test this is behaving as desired i.e. these models are classed as redundant (even with the CDS diffs)

**Comment by lucventurini** [comment]
Hi @swarbred,

Can you send me by mail the location of the data? If things are handled differently between single threading and multithreading, it might be a sign of something worrisome going on. I'd rather check now if possible.

**Comment by swarbred** [comment]
Hi @lucventurini 
It's not in a location you can access, hold off for now, I will redo the full run and check that first.

**Comment by swarbred** [comment]
So this is odd, it's not the multithreading I can reproduce the https://github.com/EI-CoreBioinformatics/mikado/issues/267#issuecomment-588119257 selection of transcripts but only when running the whole scaffold/contig through rather than the region even though the region is not excluding any transcripts from what I can see in the browser.

**PickRun5_test_5 on full contig**
```
source mikado-2.0rc6_0cfb20a_CBG; /usr/bin/time -v mikado pick -p 1 --json-conf configuration_run2_test.toml --subloci-out mikado.subloci.gff3 --monoloci-out mikado.monoloci.gff3 --output-dir ./PickRun5_test_5 -r Calendula_officinalis_EIV1.1_Contig_0000023:1..108649 -lv DEBUG --log pick.log
```
**PickRun5_test on specific region of the contig**

```
source mikado-2.0rc6_0cfb20a_CBG; /usr/bin/time -v mikado pick -p 1 --json-conf configuration_run2_test.toml --subloci-out mikado.subloci.gff3 --monoloci-out mikado.monoloci.gff3 --output-dir ./PickRun5_test -lv DEBUG -r Calendula_officinalis_EIV1.1_Contig_0000023:34580..36663 --log pick.log
```
[PickRun5_test_mikado.loci.gff3.txt](https://github.com/EI-CoreBioinformatics/mikado/files/4226796/PickRun5_test_mikado.loci.gff3.txt)
[PickRun5_test_pick.log.txt](https://github.com/EI-CoreBioinformatics/mikado/files/4226799/PickRun5_test_pick.log.txt)

[PickRun5_test_5_mikado.loci.gff3.txt](https://github.com/EI-CoreBioinformatics/mikado/files/4226800/PickRun5_test_5_mikado.loci.gff3.txt)
[PickRun5_test_5_pick.log.txt](https://github.com/EI-CoreBioinformatics/mikado/files/4226805/PickRun5_test_5_pick.log.txt)
[configuration_run2_test.toml.txt](https://github.com/EI-CoreBioinformatics/mikado/files/4226807/configuration_run2_test.toml.txt)

If you see the PickRun5_test_5 files these contain models marked as classcode =

also looking in the log file (PickRun5_test_5_pick.log.txt) for one of the redundant models (CoDisc_scallop_CoDisc_sca.427.0.2) indicates
```
DEBUG - add_transcript_to_locus - LociProcesser-1 - CoDisc_scallop_CoDisc_sca.427.0.2 is a valid splicing isoform; Ccode: =
```
where in the PickRun5_test_pick.log.txt this shows

```
DEBUG - add_transcript_to_locus - LociProcesser-1 - CoDisc_scallop_CoDisc_sca.427.0.2 not added because it is not a                 valid splicing isoform. Ccode: =
```
Why we have a difference between these runs seems odd but that may be a distraction the main issue is why we get what I would regard as redundant models i.e. as in the pic above.

**Comment by lucventurini** [comment]
@swarbred 

This is really odd. I have to dig into the logs to find out what's happening, hopefully I will get an idea soon. I will keep you posted tomorrow.

**Comment by lucventurini** [comment]
@swarbred 

Please see a2de57e. After this commit, Mikado will try really hard to find redundant splicing isoforms after padding. Hopefully this should solve the issue.

Please let me know, back to #255 now.

**Comment by swarbred** [comment]
@lucventurini 
A separate (but I assume related) issue present in 0cfb20a (PickRun5 in the screenshot) but absent from the version i have installed as 3dc2ed9 (PickRun3)

<img width="946" alt="Screen Shot 2020-02-20 at 15 05 46" src="https://user-images.githubusercontent.com/3267809/74947391-af82d700-53f2-11ea-979b-84eedc9eea31.png">

only one transcript but as you can see this has been padded in PickRun5 (based on the transcript below)

As before testing this on a specific region (test6)
```
source mikado-2.0rc6_0cfb20a_CBG; /usr/bin/time -v mikado pick -p 1 --json-conf configuration_run2_test.toml --subloci-out mikado.subloci.gff3 --monoloci-out mikado.monoloci.gff3 --output-dir ./PickRun5_test6 -lv DEBUG -r Calendula_officinalis_EIV1.1_Contig_0000019:109217..111776
```
and the output looks fine (not padded)
[PickRun5_test6_out_swarbred.mikado-2.0rc6_0cfb20a_CBG.configuration_run2_test.25814169.err.txt](https://github.com/EI-CoreBioinformatics/mikado/files/4231563/PickRun5_test6_out_swarbred.mikado-2.0rc6_0cfb20a_CBG.configuration_run2_test.25814169.err.txt)
[PickRun5_test6.mikado.loci.gff3.txt](https://github.com/EI-CoreBioinformatics/mikado/files/4231564/PickRun5_test6.mikado.loci.gff3.txt)

with the full scaffold this is getting padded
```
source mikado-2.0rc6_0cfb20a_CBG; /usr/bin/time -v mikado pick -p 1 --json-conf configuration_run2_test.toml --subloci-out mikado.subloci.gff3 --monoloci-out mikado.monoloci.gff3 --output-dir ./PickRun5_test7 -lv DEBUG -r Calendula_officinalis_EIV1.1_Contig_0000019:1..113843
```

[PickRun5_test7_mikado.loci.gff3.txt](https://github.com/EI-CoreBioinformatics/mikado/files/4231571/PickRun5_test7_mikado.loci.gff3.txt)
[PickRun5_test7_out_swarbred.mikado-2.0rc6_0cfb20a_CBG.configuration_run2_test.25814208.err.txt](https://github.com/EI-CoreBioinformatics/mikado/files/4231580/PickRun5_test7_out_swarbred.mikado-2.0rc6_0cfb20a_CBG.configuration_run2_test.25814208.err.txt)

**Comment by swarbred** [comment]
@lucventurini I take it we hope a2de57e fixes the above issue as well?

why are we not seeing the issue when running on a smaller region (makes testing this take a little longer)

**Comment by lucventurini** [comment]
> why are we not seeing the issue when running on a smaller region (makes testing this take a little longer)

I am not sure about why, to be honest :-( I know what caused the bug, but I cannot understand why it happens in one case but not the other.
It might be due to the transcripts that are initially added to the region, maybe? As in, some transcripts are excluded when you extract the region, and they would trigger the bug?

**Comment by swarbred** [comment]
> It might be due to the transcripts that are initially added to the region, maybe? As in, some transcripts are excluded when you extract the region, and they would trigger the bug?

That was my original thought but I don't see the evidence for this, this is all the transcripts that we are using and non would be excluded based on the region selected


<img width="998" alt="Screen Shot 2020-02-20 at 15 54 05" src="https://user-images.githubusercontent.com/3267809/74954858-acd8af80-53fb-11ea-9d3c-125ae5c4b4c3.png">


I've also checked the log and all unique transcripts are present

```
Submitting locus # 1 (Calendula_officinalis_EIV1.1_Contig_0000019:109746-111096), with transcripts:
CoLeaf_scallop_CoLeaf_sca.334.0.0,CoLeaf_stringtie_CoLeaf_str.319.1,CoLeaf.hq_lq_UnnamedSample_HQ_transcript/64500.mrna1,CoDisc_scallop_CoDisc_sca.370.0.0,CoRay_scallop_CoRay_sca.327.0.0
```
Something is different when running on the smaller region but I dont see what this could be

**Comment by lucventurini** [comment]
> Something is different when running on the smaller region but I dont see what this could be

I ... honestly do not know. A cursory look at the code does not reveal anything obvious. I hope that the new commit will solve the issue nonetheless, but I am puzzled by this behaviour.

I will have to look at the log to have a hope to understand what's going on.

**Comment by swarbred** [comment]
Again agree both the examples are now correct

**Comment by swarbred** [comment]
@lucventurini 
Based on my full runs, I consider this resolved in d094f99

**Comment by lucventurini** [comment]
> @lucventurini
> Based on my full runs, I consider this resolved in [d094f99](https://github.com/EI-CoreBioinformatics/mikado/commit/d094f995c4a609fac2c117655a5044f56258dd5e)

Great news! Does this also solve #266?
